docs/Toy/toy/dialect.py

```diff
@@ -0,0 +1,267 @@
+        return super().create(operands=[input])
+
+
+@irdl_op_definition
+class ReturnOp(Operation):
+    """
+    The "return" operation represents a return operation within a function.
+    The operation takes an optional tensor operand and produces no results.
+    The operand type must match the signature of the function that contains
+    the operation. For example:
+
+    ```mlir
+      func @foo() -> tensor<2xf64> {
+        ...
+        toy.return %0 : tensor<2xf64>
+      }
+    ```
+    """
+    name: str = 'toy.return'
+    arguments: Annotated[OptOperand, AnyTensorTypeF64]
+
+    @classmethod
+    def from_input(cls: type[ReturnOp],
+                   input: Optional[SSAValue] = None) -> ReturnOp:
+        return super().create(operands=[input] if input is not None else [])
+
+
+@irdl_op_definition
+class ReshapeOp(Operation):
+    """
+    Reshape operation is transforming its input tensor into a new tensor with
+    the same number of elements but different shapes. For example:
+
+    ```mlir
+       %0 = toy.reshape (%arg1 : tensor<10xf64>) to tensor<5x2xf64>
+    ```
+    """
+    name: str = 'toy.reshape'
+    arguments: Annotated[VarOperand, AnyTensorTypeF64]
+    # We expect that the reshape operation returns a statically shaped tensor.
+    res: Annotated[OpResult, TensorTypeF64]
+
+    @classmethod
+    def from_input(cls: type[ReshapeOp], input: SSAValue,
+                   shape: List[int]) -> ReshapeOp:
+        assert isinstance(input.typ, TensorType | UnrankedTensorType)
+        element_type = cast(Float64Type,
+                            cast(TensorType[Any], input.typ).element_type)
+        t = AnyTensorType.from_type_and_list(element_type, shape)
+        return super().create(result_types=[t], operands=[input])
+
+    def verify_(self):
+        result_type = self.res.typ
+        assert isinstance(result_type, TensorType)
+        result_type = cast(TensorTypeF64, result_type)
+        if not len(result_type.shape.data):
+            raise VerifyException(
+                'Reshape operation result shape should be defined')
+
+
+@irdl_op_definition
+class TransposeOp(Operation):
+    name: str = 'toy.transpose'
+    arguments: Annotated[Operand, AnyTensorTypeF64]
+    res: Annotated[OpResult, AnyTensorTypeF64]
+
+    @staticmethod
+    def from_input(input: SSAValue):
+        input_type = input.typ
+        assert isinstance(input_type, TensorType | UnrankedTensorType)
+        output_type: TensorType[Any] | UnrankedTensorType[Any]
+        if isinstance(input_type, TensorType):
+            element_type = cast(Float64Type,
+                                cast(TensorType[Any], input_type).element_type)
+            output_type = TensorType.from_type_and_list(
+                element_type, list(reversed(input_type.shape.data)))
+        elif isinstance(input_type, UnrankedTensorType):
+            output_type = input_type
+        else:
+            assert False, f'{input_type}: {type(input_type)}'
```

Comment:
Why do you assert like this and not raise an exception?

Reply:
technically this is unreachable code because of the assert above :)

Reply:
It just feels weird to see assertions with error prints. If it is unreachable, I would just leave out the else...

Reply:
I agree, although I see `assert False` as unreachability markers that make the rest of the code more readable

Reply:
A bit of a special case